pkg/image/prune/controller.go

```diff
@@ -0,0 +1,185 @@
+
+func (c *PruningController) addImagesToGraph(g graph.Graph, images *imageapi.ImageList) {
+	for i := range images.Items {
+		image := &images.Items[i]
+		if !strings.HasPrefix(image.DockerImageReference, c.registryURL) {
+			continue
+		}
+		age := util.Now().Sub(image.CreationTimestamp.Time)
+		if age.Minutes() < minimumImageAgeInMinutesToPrune {
+			continue
+		}
+		graph.Image(g, image)
+	}
+}
+
+func (c *PruningController) addImageStreamsToGraph(g graph.Graph, streams *imageapi.ImageStreamList) {
+	for i := range streams.Items {
+		stream := &streams.Items[i]
+		if len(stream.Spec.DockerImageRepository) > 0 {
+			// skip streams that point at external registries
+			continue
+		}
+		imageStreamNode := graph.ImageStream(g, stream)
+		for _, history := range stream.Status.Tags {
+			if len(history.Items) == 0 {
+				continue
+			}
+			for i := range history.Items {
+				glog.Infof("Looking for node with image id %q", history.Items[i].Image)
+				imageNode := graph.FindImage(g, history.Items[i].Image)
+				glog.Infof("Found: %#v", imageNode)
+				if imageNode != nil {
+					var kind int
+					switch i {
+					case 0:
+						kind = graph.ReferencedImageGraphEdgeKind
+					default:
+						kind = graph.WeakReferencedImageGraphEdgeKind
+					}
+					g.AddEdge(imageStreamNode, imageNode, kind)
+				}
+			}
+		}
+	}
+}
+
+func (c *PruningController) addPodsToGraph(g graph.Graph, pods *kapi.PodList) {
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+		if pod.Status.Phase != kapi.PodRunning && pod.Status.Phase != kapi.PodPending {
+			// not pending or running, skip
+			continue
+		}
+
+		podNode := graph.Pod(g, pod)
+		c.addPodSpecToGraph(g, &pod.Spec, podNode)
+	}
+}
+
+func (c *PruningController) addPodSpecToGraph(g graph.Graph, spec *kapi.PodSpec, predecessor gonum.Node) {
+	for j := range spec.Containers {
+		container := spec.Containers[j]
+		if !strings.HasPrefix(container.Image, c.registryURL) {
+			continue
+		}
+		parts := strings.Split(container.Image, "@")
+		if len(parts) != 2 {
+			continue
+		}
+		id := parts[1]
+		if len(id) == 0 {
+			continue
+		}
+		glog.Infof("Looking for node with image id %q", id)
+		imageNode := graph.FindImage(g, id)
+		glog.Infof("Found: %#v", imageNode)
+		if imageNode != nil {
+			g.AddEdge(predecessor, imageNode, graph.ReferencedImageGraphEdgeKind)
+		}
+	}
+}
+
+func (c *PruningController) addReplicationControllersToGraph(g graph.Graph, rcs *kapi.ReplicationControllerList) {
+	for i := range rcs.Items {
+		rc := &rcs.Items[i]
+		rcNode := graph.ReplicationController(g, rc)
+		c.addPodSpecToGraph(g, &rc.Spec.Template.Spec, rcNode)
+	}
+}
+
+func (c *PruningController) prune(g graph.Graph) error {
+	//prunable := []*graph.ImageNode{}
+
+	// TODO not efficient
+	allNodes := g.NodeList()
+	for i := range allNodes {
+		node := allNodes[i]
+		imageNode, ok := node.(*graph.ImageNode)
+		if !ok {
+			continue
+		}
+
+		onlyWeakReferences := true
+		imageStreamPredecessors := []*imageapi.ImageStream{}
```

Comment:
Hrm - would prefer that you find a way to keep the coupling low between various parts here.  Doing a second graph traversal isn't that bad.

----- Original Message -----
> > +}
> > +
> > +func (c *PruningController) prune(g graph.Graph) error {
> > +	//prunable := []*graph.ImageNode{}
> > +
> > +	// TODO not efficient
> > +	allNodes := g.NodeList()
> > +	for i := range allNodes {
> > +		node := allNodes[i]
> > +		imageNode, ok := node.(*graph.ImageNode)
> > +		if !ok {
> > +			continue
> > +		}
> > +
> > +		onlyWeakReferences := true
> > +		imageStreamPredecessors := []*imageapi.ImageStream{}
> 
> I have more locally - it's to update stream.status.tags to remove images
> deleted by the pruning.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1841/files#r28805810